### ISSUE TITLE
Lazy module import into variable with union type causes wrong type error

### ISSUE THREAD

**Issue by MrWolfZ** [microsoft/TypeScript#48585]
# Bug Report

### 🔎 Search Terms

lazy, import

### 🕗 Version & Regression Information

I observed this first in version 4.6.2. I then verified that the issue still exists in the latest nightly build (v4.7.0-dev.20220406) and at least as far back as version 3.9.7 (it likely existed even before that but I think checking the latest 3.X version should be enough).

### ⏯ Code & Bug Workbench Link

```ts repro
// @filename: one.ts
export const one = 1

// @filename: two.ts
const x: Promise<1> = import('./one').then((m) => m.one)
const y: Promise<1> | Promise<2> = import('./one').then((m) => m.one) // wrong type error here
const z: Promise<1> | Promise<2> = x

```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?ts=4.7.0-dev.20220406#code/PTAEAEDMEsBsFMB2BDAtvAXKA9o+A6AFwGcAoeADwAdsAnQ0AY12Id3lAF5QBGU0kBBgIU6LIQDu2ImWaJWoClgAKtbKmjF4AHh4A+LqGioa9ABQByfMHYWAlEQAWSM2dR2uB1PnZ3SchQBPFTUNLV0DAB9QVXVNHQAmA25jU0JLa1sHQmdEV3dPUG9ffxYGAC8QuPD9UGjYsMTkxVIgA)

### 🙁 Actual behavior

In the code sample above, the assignment to `y` causes a wrong type error, while assigning to `z` from an intermediate variable does not cause a type error. I don't think I need to explain why this is wrong.

The error also seems to indicate that type inference was somehow done in the wrong direction, since it seems to think that the type of the import is `Promise<1 | 2>` while it clearly is `Promise<1>`:

```
Type 'Promise<1 | 2>' is not assignable to type 'Promise<1> | Promise<2>'.
  Type 'Promise<1 | 2>' is not assignable to type 'Promise<1>'.
    Type '1 | 2' is not assignable to type '1'.
      Type '2' is not assignable to type '1'.(2322)
```

### 🙂 Expected behavior

The assignment to `y` should not cause a type error.

**Comment by RyanCavanaugh** [comment]
Modules not necessary:
```ts
declare const onePromise: Promise<{ one: 1 }>;

const x: Promise<1> = onePromise.then((m) => m.one)
const y: Promise<1> | Promise<2> = onePromise.then((m) => m.one)
const z: Promise<1> | Promise<2> = x
```
The problem here is that when finding inference candidates for `T`'s type parameter, we see both `1` and `2`, and those have equal footing with the other candidates. For other types this is generally desirable, but `Promise` a bit special due to its callback nature and this ends up doing a bit more harm than good.

The "fix" is to write a type argument
```
const y: Promise<1> | Promise<2> = onePromise.then<1>((m) => m.one)
```

**Comment by MrWolfZ** [comment]
This behavior is a bit unfortunate since it affects a public API method of a library I'm working on and I don't want to force my users to have to specify the types, but I guess I'll find a workaround.
